Use filePath instead of fileName when running Abaqus with script and change `tests`

Diogo-Rossi opened the pull request:
# Description

Fixes #1103, detailed in discussion #1095

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Test A: `test_mdb.py`
- [x] Test B: `test_odb.py`

The tests were modified to accept 3 situations:
1. The current work directory (CWD) is `temp`
2. The CWD is `tests`
3. The CWD is the root folder

* In each situation, the test scripts adds the `src` folder to `PATH` and change the CWD to `temp` folder to run Abaqus comand

* The test scripts needs also to change the value of `sys.argv[0]` to the script's path (`__file__`) when it is executed from `pytest`, because the Abaqus command needs to use this value, and it is set to another when using `pytest`.

* The `test_odb.py` file also needs to define a `pytest` class when it is running in Abaqus interpreter, because there is no `pytest` module in this interpreter.

**Test Configuration**:
* System version: Windows 10
* Abaqus version: 2021
* Python version: 3.9
* abqpy version: 2021

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

haiiliin commented:
Do we have to remove the temp folder in the tests?

haiiliin commented:
The `openOdb` function is also included in the `odbAccess` and `visualization` modules, can you modify these functions too? The `openOdb` function should be imported from `abaqus.Odb.OdbCommands` we can modify `abaqus.Odb.OdbCommands.openOdb` and import it in `odbAccess` and `visualization`.

Diogo-Rossi commented:
> Do we have to remove the temp folder in the tests?

No, I added the `temp` folder to run the Abaqus simulations in it, and I modified the tests to change the CWD to it.
Then, the Abaqus temporary files are kept in this folder and are always overwritten.

This modification also servers as a test for the new feature. But, if you prefer to kept the previous format, I can remove these modification in the tests.

haiiliin commented:
> Abaqus simulations in it, and I modified the tests to change the CWD to it.
> Then, the Abaqus temporary files are kept in this folder and are always overwritten.

How about creating the `temp` folder dynamically when executing pytests?

Diogo-Rossi commented:
> > Abaqus simulations in it, and I modified the tests to change the CWD to it.
> > Then, the Abaqus temporary files are kept in this folder and are always overwritten.
> 
> How about creating the `temp` folder dynamically when executing pytests?

That is a possible approach.
But then, we will have check if the `temp` folder exists before try to create it in every test, because the tests are run in parallel. Also, we will have to delete this folder at the end of the test, and that could cause conflict with others running tests.
So, because of the possible conflicts and the need to write many additional lines, I think it is better to keep this folder empty with the `.gitkeep` file to use with `pytest`. What do you think?

haiiliin commented:
> * The `test_odb.py` file also needs to define a `pytest` class when it is running in Abaqus interpreter, because there is no `pytest` module in this interpreter.

Why is this necessary? The tests will only be used for testing and they shouldn't be submitted to Abaqus.

haiiliin commented:
> tests are run in parallel

Will the tests be run in parallel? By default, the `pytests` should be run in sequential order

Diogo-Rossi commented:
> > * The `test_odb.py` file also needs to define a `pytest` class when it is running in Abaqus interpreter, because there is no `pytest` module in this interpreter.
> 
> Why is this necessary? The tests will only be used for testing and they shouldn't be submitted to Abaqus.

Oh, I thought that the tests should be submitted to Abaqus, that is the reason of theses modifications in the tests files.

Diogo-Rossi commented:
> > tests are run in parallel
> 
> Will the tests be run in parallel? By default, the `pytests` should be run in sequential order

Oh, I thought that the tests were run in parallel, my mistake. But then we would need to know which test runs first to create `temp` folder,  and which runs last, to delete it. I still think it would be better to keep the folder.

But, if the tests [shouldn't be submitted to Abaqus](https://github.com/haiiliin/abqpy/pull/1104#issuecomment-1240280944), we don't need this folder anymore.

haiiliin commented on tests/test_mdb.py:
```diff
@@ -1,5 +1,23 @@
-import os, sys
-sys.path.insert(0,os.path.abspath('./src'))
+import sys, os
+# Execute only if in Python 3.x interpreter
+if sys.version_info[0] > 2:
+    # Check which folder the test are running from and then
+    # add the `src` folder to PATH for testing. Also change
+    # the work directory to `temp` to run Abaqus command in it.
+    folder = os.path.basename(os.getcwd())
+    if folder == 'temp':
+        sys.path.insert(0,os.path.abspath('../../src'))
+    elif folder == 'tests':
+        sys.path.insert(0,os.path.abspath('../src'))
+        os.chdir('temp')
+    else:
+        sys.path.insert(0,os.path.abspath('./src'))
+        os.chdir('tests/temp')
+    
+    #This check is needed to run from `pytest` framework
+    if sys.argv[0] != __file__:
+        sys.argv.insert(0,__file__)
```
This is quite complicated, how about use the `pytest_configure` function in the `conftest.py` file (https://stackoverflow.com/a/53690092)? 
```python
def pytest_configure(config):
    os.system('python setup.py install')
```

Diogo-Rossi replied:
Do you mean, include that block of code in `conftest.py` inside `pytest_configure`?

Diogo-Rossi replied:
But, if we don't need to submit the tests to Abaqus, we won't need this block of code either

haiiliin replied:
If we do not install the `abqpy` package locally, the `pytest` will fail

haiiliin replied:
In the GitHub workflows, the `abqpy` will always be installed automatically, in case we have to run `pytest` locally and sometimes we forgot to install it, we might need this.

Diogo-Rossi replied:
> In the GitHub workflows, the `abqpy` will always be installed automatically, in case we have to run `pytest` locally and sometimes we forgot to install it, we might need this.
Ok, but we won't need the `os.chdir('temp')` anymore, right?

haiiliin replied:
yes

Diogo-Rossi replied:
Done. Removed the modifications in [642d0e2](https://github.com/haiiliin/abqpy/pull/1104/commits/642d0e26c38e8eca004f4af6bba3ac8628436d10). Now the tests doesn't submit to Abaqus.

haiiliin commented:
> > > tests are run in parallel
> > 
> > 
> > Will the tests be run in parallel? By default, the `pytests` should be run in sequential order
> 
> Oh, I thought that the tests were run in parallel, my mistake. But then we would need to know which test runs first to create `temp` folder, and which runs last, to delete it. I still think it would be better to keep the folder.
> 
> But, if the tests [shouldn't be submitted to Abaqus](https://github.com/haiiliin/abqpy/pull/1104#issuecomment-1240280944), we don't need this folder anymore.

Yes, it is just for testing, I think we don't even need to run it in the `temp` folder.

Diogo-Rossi commented:
> The `openOdb` function is also included in the `odbAccess` and `visualization` modules, can you modify these functions too? The `openOdb` function should be imported from `abaqus.Odb.OdbCommands` we can modify `abaqus.Odb.OdbCommands.openOdb` and import it in `odbAccess` and `visualization`.

I was thinking about that and I guessed some different modification in `odbAccess.openOdb` function.

Diogo-Rossi commented:
The [`odbAccess.openOdb`](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=9f0ec3921d15431997232cb3f25d285d#simaker-odbodbopenodbpyc) function is different from the [`session.openOdb`](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=d16c67175beb4d0aa7584a7d4f2bfd5a#simaker-odbodbsesopenodbpyc).

The `odbAccess.openOdb` function can be run regardless of Abaqus CAE. See, for instance [this example](https://help.3ds.com/2022/english/dssimulia_established/SIMACAECMDRefMap/simacmd-c-odbintrosimpleexapyc.htm?contextscope=all&id=f32d4a1eda50408b96561de06760f215).
Its usage is simply `abaqus python odbRead.py`, so I think that the command in this function could be changed to the following:
```python
os.system(f"{abaqus} python {filePath}")
```
What do you think?

haiiliin commented:
> 

But following [this](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=9f0ec3921d15431997232cb3f25d285d#simaker-odbodbopenodbpyc), is seems `odbAccess.openOdb` function just calls the `session.openOdb` method

haiiliin commented:
> But following [this](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=9f0ec3921d15431997232cb3f25d285d#simaker-odbodbopenodbpyc), is seems that `odbAccess.openOdb` function just calls the `session.openOdb` method

Oh, I made a mistake, I am looking at the `session.openOdb` method

Diogo-Rossi commented:
> The `openOdb` function is also included in the `odbAccess` and `visualization` modules, can you modify these functions too? The `openOdb` function should be imported from `abaqus.Odb.OdbCommands` we can modify `abaqus.Odb.OdbCommands.openOdb` and import it in `odbAccess` and `visualization`.

Done in [bc2fe1e](https://github.com/haiiliin/abqpy/pull/1104/commits/bc2fe1ef838c6ca5496c5e6b36e2c725336ab22a)

Diogo-Rossi commented:
In [b3cb432](https://github.com/haiiliin/abqpy/pull/1104/commits/b3cb4326932a7c48af6b40a18b43319bc0031c98), I modified the Abaqus command inside `odbAccess.openOdb` to not use Abaqus CAE anymore. i.e.:
```python
os.system(f"{abaqus} python {filePath}")
```
Which is the intention of `odbAccess.openOdb` method inside some scripts. See, for instance, the example script obtained with
```cmd
abaqus fetch job=odbRead
```
The top of the file is:
```python
# odbRead.py
# A script to read data from an Abaqus odb file 
# The Script is using the odbAccess module 
# Output database is opened and displacement data 
# from the node at the center of the hemispherical punch.
# Usage : 
# abaqus python odbRead.py
```

Diogo-Rossi commented:
> If we do not provide the Odb `database` file like `session.openOdb`, it is still valid? When we use the `odbAccess.openOdb` in the example output script file, the result is still correct?

Yes, the database is informed inside the script, in the second line of it.

```python
odb = openOdb(path='viewer_tutorial.odb')
```

Diogo-Rossi commented:
> If we do not provide the `.odb database` file like `session.openOdb`, it is still valid? When we use the `odbAccess.openOdb` in the example output script file, the result is still correct?

Actually, it is a question of mine. Why the Abaqus command inside `session.openOdb` isn't the command bellow?
```python
os.system(f"{abaqus} cae -noGUI {filePath}")
```
The database would normally be accessed inside abaqus with the line `odb = session.openOdb('Job-1.odb')`, wouldn't it?

haiiliin commented:
> 

I mean in `abaqus python`, we don't provide the `database`explicitly, when using `abaqus cae` without providing `database` explicitly, we can't get the output.

Diogo-Rossi commented:
> > 
> 
> I mean in `abaqus python`, we don't provide the `database`explicitly, when using `abaqus cae` without providing `database` explicitly, we can't get the output really.

I will make some tests to check it out.

haiiliin commented:
> > If we do not provide the `.odb database` file like `session.openOdb`, it is still valid? When we use the `odbAccess.openOdb` in the example output script file, the result is still correct?
> 
> Actually, it is a question of mine. Why the Abaqus command inside `session.openOdb` isn't the command bellow?
> 
> ```python
> os.system(f"{abaqus} cae -noGUI {filePath}")
> ```
> 
> The database would normally be accessed inside abaqus with the line `odb = session.openOdb('Job-1.odb')`, wouldn't it?

I tried this initially, didn't know why, and then I use the later one.

Diogo-Rossi commented:
> I tried this initially, didn't work, didn't know why, and then I use the later one. And we can't get the output using `-noGUI` option.

Yes, I made the wrong question. The line I was talking about is
```python
os.system(f"{abaqus} cae script={filePath}")
```
Without the `-noGUI` option.

I've made a test and got the output, but I needed to add the following lines after `session.openOdb`:
```python
o = session.odbs['Job-1.odb']
session.viewports['Viewport: 1'].setValues(displayedObject=o)
```

Then, the original whole test code becomes:

```python
from abaqus import *
from abaqusConstants import *
from driverUtils import *
import numpy as np

executeOnCaeStartup()

# Open output database
odb = session.openOdb('Job-1.odb')

o = session.odbs['Job-1.odb']
session.viewports['Viewport: 1'].setValues(displayedObject=o)

# Extract output data
dataList = session.xyDataListFromField(odb=odb, outputPosition=NODAL,
                                       variable=(('U', NODAL, ((COMPONENT, 'U3'),)),),
                                       nodeSets=('INSTANCE.SET-TOP', ))
data = np.array(dataList[0])
np.savetxt('data.csv', data, header='time,U3', delimiter=',', comments='')
```
And I don't need to provide the database in the Abaqus command.
Without those 2 additional lines, there is a error and we can't get the output.

haiiliin commented:
Oh, I think this is the right solution, we can't get the output just because the odb is not showed in the current viewport. 

By the way, I think it can be simplified to `session.viewports['Viewport: 1'].setValues(displayedObject=odb)`

haiiliin commented:
Have you tested the second one? will it produce correct data file?

Diogo-Rossi commented:
> Have you tested the second one? will it produce correct data file?

Yes, I tested it. It produces the same `data.csv` file as the one that calls Abaqus CAE

haiiliin commented:
> > Have you tested the second one? will it produce correct data file?
> 
> Yes, I tested it. It produces the same `data.csv` file as the one that calls Abaqus CAE

What if we run this script in GUI?

Diogo-Rossi commented:
> What if we run this script in GUI mode?

I've just tested it in GUI mode now and it gave the same result

But in GUI mode it opens Abaqus CAE graphical interface, which is not necessary for this script, because it does not use the Abaqus kernel process.

haiiliin commented:
So in short, for `session.openOdb`, we have to

- Add a new line in the example script to display the odb object
- Update the corresponding tutorial in the documentation


For the second one, 

- Use `abaqus python` to submit the job
- Test the script outside/inside the abaus kernel
- Update `odbAcces` and `visualization` module

haiiliin commented:
> > What if we run this script in GUI mode?
> 
> I've just tested it in GUI mode now and it gave the same result
> 
> But in GUI mode it opens Abaqus CAE graphical interface, which is not necessary for this script, because it does not use the Abaqus kernel process.
> 

Fantastic, let's make this change

Diogo-Rossi commented:
>For the first one, we have to modify the example script, add a new line to display the object, right? And also update the tutorials in the documentation.

Yes, that is correct.

> For the second one, have you tested the second one? will it produce correct data file? The `visualization` module also has a `openOdb` function, which I think it is the same with the one in `odbAccess`.

Yes, I tested it. I think that the `openOdb` function inside `visualization` module is not the same as in `odbAccess` but the one used by the `session` object, look:
* [Here is the `odbAccess` function](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=9f0ec3921d15431997232cb3f25d285d#simaker-odbodbopenodbpyc)
* [Here is the `visualization` import before use the `session.openOdb` method](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEKERRefMap/simaker-c-odbopenodbcommandpyc.htm?contextscope=all&id=d16c67175beb4d0aa7584a7d4f2bfd5a#simaker-odbodbsesopenodbpyc)

> So in short, for `session.openOdb`, we have to
> 
> * Add a new line in the example script to display the odb object
> * Update the corresponding tutorial in the documentation
> 
> For the second one,
> 
> * Use `abaqus python` to submit the job
> * Test the script outside/inside the abaus kernel
> * Update `odbAcces` and `visualization` module

Yes, that is correct.

haiiliin commented:
> Yes, I tested it. I think that the openOdb function inside visualization module is not the same as in odbAccess but the one used by the session object, look:

Right.

haiiliin commented:
This solves the problem that bothers me long time ago, we could now extract the output data in a batch file without opening the GUI interface, which is very useful when we have to do the parametric stufy.

Diogo-Rossi commented:
> So in short, for `session.openOdb`, we have to
> 
> * Add a new line in the example script to display the odb object

Done in [f83b392](https://github.com/haiiliin/abqpy/pull/1104/commits/f83b392946266643a07c5562b2ef23a7c12c3d34) and [015ff24](https://github.com/haiiliin/abqpy/pull/1104/commits/015ff2421f52c231b840a6a00c4747019add384f)
> * Update the corresponding tutorial in the documentation

Done in [855e9a3](https://github.com/haiiliin/abqpy/pull/1104/commits/855e9a37faaf91e0c0a1b346d9a42b22757bcd72)
> 
> For the second one,
> 
> * Use `abaqus python` to submit the job

Already done in [b3cb432](https://github.com/haiiliin/abqpy/pull/1104/commits/b3cb4326932a7c48af6b40a18b43319bc0031c98)
> * Test the script outside/inside the abaus kernel

Did the tests locally.
> * Update `odbAcces` and `visualization` module

Already done in [b3cb432](https://github.com/haiiliin/abqpy/pull/1104/commits/b3cb4326932a7c48af6b40a18b43319bc0031c98) for `odbAcces`. `visualization` module remains the same.

haiiliin commented:
How about the output scripts in the example folder?

Diogo-Rossi commented:
> How about the output scripts in the example folder?

I wasn't aware. I'll do that now.

haiiliin commented:
Ok, is there anything else? Otherwise, I will merge this.

haiiliin commented:
> > How about the output scripts in the example folder?
> 
> I wasn't aware. I'll do that now.

Also the Jupyter output script

Diogo-Rossi commented:
> Ok, is there anything else? Otherwise, I will merge this.

No, I think that is all.

I had some new idea, but will put it on discussion first.

Diogo-Rossi commented:
> > > How about the output scripts in the example folder?
> > 
> > 
> > I wasn't aware. I'll do that now.
> 
> Also the Jupyter output script

Oh, sorry. I'll do that now

haiiliin commented:
Now it is ready to merge, right?

Diogo-Rossi commented:
> Now it is ready to merge, right?

I think so.

haiiliin commented:
Ok, I will enable the auto-merge now.
